File: Blind75-Practice/easy/02-sliding-window/arrays/sliding-window-maximum.py
"""
239. Sliding Window Maximum - https://leetcode.com/problems/sliding-window-maximum/
Difficulty: Hard
Pattern: Sliding Window + Deque

You are given an array of integers nums, there is a sliding window of size k which is moving from the very left of the array to the very right. You can only see the k numbers in the window. Each time the sliding window moves right by one position.

Return the max sliding window.

Example 1:
Input: nums = [1,3,-1,-3,5,3,6,7], k = 3
Output: [3,3,5,5,6,7]

Example 2:
Input: nums = [1], k = 1
Output: [1]

Time: O(n), Space: O(k)
"""

from typing import List
from collections import deque


def max_sliding_window(nums: List[int], k: int) -> List[int]:
    """
    Sliding window with deque to maintain max elements
    Time: O(n), Space: O(k)
    """
    # TODO: Implement deque solution
    pass


# Test cases
if __name__ == "__main__":
    # Test 1
    nums1, k1 = [1, 3, -1, -3, 5, 3, 6, 7], 3
    print(f"Max Sliding Window: {max_sliding_window(nums1, k1)}")  # [3,3,5,5,6,7]

    # Test 2
    nums2, k2 = [1], 1
    print(f"Max Sliding Window: {max_sliding_window(nums2, k2)}")  # [1]

File: Blind75-Practice/hard/05-sliding-window/Java/SlidingWindowMaximum.java
/*
 * 239. Sliding Window Maximum
 * Difficulty: Hard
 * Pattern: Sliding Window / Monotonic Deque
 *
 * Problem:
 * You are given an array of integers `nums`, and there is a sliding window of size `k`
 * moving from the very left of the array to the very right. You can only see the `k`
 * numbers inside the window at a time.
 *
 * Return the maximum value in each sliding window.
 *
 * Example 1:
 * Input: nums = [1,3,-1,-3,5,3,6,7], k = 3
 * Output: [3,3,5,5,6,7]
 *
 * Example 2:
 * Input: nums = [1], k = 1
 * Output: [1]
 *
 * Constraints:
 * - 1 <= nums.length <= 10^5
 * - -10^4 <= nums[i] <= 10^4
 * - 1 <= k <= nums.length
 *
 */

import java.util.Arrays;

public class SlidingWindowMaximum {
    public int[] maxSlidingWindow(int[] nums, int k) {

    // return new int[stack.size] { turn stack into int array}
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    // 
    //

    // int n = nums.length;
    // int[] answer = new int[n - k + 1];
    // Deque<Integer> deque = new ArrayDeque<>();
    // for (int i = 0; i < n; i++) {
    //     while (!deque.isEmpty() && deque.peekFirst() <= i - k) {
    //         deque.pollFirst();
    //     }
        
    //     while (!deque.isEmpty() && nums[deque.peekLast()] <= nums[i]) {
    //         deque.pollLast();
    //     }

    //     deque.offerLast(i);

    //     if (i >= k - 1) {
    //         answer[i - k + 1] = nums[deque.peekFirst()];
    //     }
    // }
    // return answer;
    }
    

    public static void main(String[] args) {
        SlidingWindowMaximum sol = new SlidingWindowMaximum();
        assert Arrays.equals(sol.maxSlidingWindow(new int[] {1, 3, -1, -3, 5, 3, 6, 7}, 3), new int[] {3, 3, 5, 5, 6, 7});
        assert Arrays.equals(sol.maxSlidingWindow(new int[] {1}, 1), new int[] {1});
        System.out.println("All test cases passed!");
    }
}
